### PR TITLE
CT-2862: upgrade rails to 5.0

### PR DESCRIPTION
**Pull request by ymao2**
Upgrade app from rails 4.2 to 5.0 - another PR will upgrade to 5.2

### Notes

There are a lot of changes here, so selected the following non-trivial commits are ones I feel are important to discuss:

b5e7f42bc37731af2f8033f555a260bef062fbdf - should Puma be here

d568fa514a147eda37cce51dbaec0f6823b47617 - removed zeros in sql, any issues? I
dan thing of any.

f85f44cfd9d3ea768b520e3e03933b43cff24dd7 - rails 4 UTF-8 issue

3ff3f07d32214bd4bdd2bad532f433fc4088e0bf - be_visible rather than be_all_there

a9121c331793af57c2a09393eed2b229388d783e - person.reload issue
c9bd0c4a615b737482f6feeba7d2fae8ee05db35

2d1584f9b27c2e5bdb292a40f296ade115fcb0c5 - touch operations and nil versions
audit trail config

2a576bb46dcb1ae816a42e679952c434087a0a27 - `redirect_to :back` is deprecated.
Is home path correct.

3d178dce5702674c469e4aa085dcb8361896c079 - ping_controller and
health_check_controller inheritance opt outs

### CODE REVIEW

File: spec/controllers/search_controller_spec.rb [hunk 1]
```diff
@@ -47,27 +47,27 @@
...
-      it 'assigns team search result to @team_results' do
-        subject
-        expect(assigns(:team_results)).to eq(team_results)
-      end
-    end
+    #   it 'assigns team search result to @team_results' do
+    #     subject
+    #     expect(assigns(:team_results)).to eq(team_results)
+    #   end
+    # end
```
Comment (Polsonby) [inline spec/controllers/search_controller_spec.rb:68]: Maybe let's add a TODO for the commented out spec and create a tech debt ticket to return to it subsequently? I think you suggested maybe we'd have another look at this after the 5.0->5.2 and if it doesn't work then, maybe we can consider deleting the spec at that point.

File: spec/features/audit_trail_spec.rb [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+describe 'Audit trail' do
   include PermittedDomainHelper
 
   let(:person) { create(:super_admin, email: 'test.user@digital.justice.gov.uk') }
   before do
     token_log_in_as person.email
   end
 
-  scenario 'Auditing an edit of a person' do
+  it 'Auditing an edit of a person' do
```
Comment (Polsonby) [inline spec/features/audit_trail_spec.rb:11]: Just interested! What's the background to this change and the one above? feature/describe and scenario/it...

File: spec/services/group_lister_spec.rb [hunk 3]
```diff
@@ -28,8 +28,8 @@
     end
   end
 
-  context 'a node with a hierarchy' do
-    subject { described_class.new.list.sort_by(&:id).last }
+  context 'with a node with a hierarchy' do
+    subject { described_class.new.list.max_by(&:id) }
```
Comment (Polsonby) [inline spec/services/group_lister_spec.rb:32]: nice, stealing this! Never seen it before...  :)

File: spec/teaspoon_env.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Set RAILS_ROOT and load the environment if it's not already loaded.
 unless defined?(Rails)
-  ENV["RAILS_ROOT"] = File.expand_path("../../", __FILE__)
-  require File.expand_path("../../config/environment", __FILE__)
+  ENV["RAILS_ROOT"] = File.expand_path('..', __dir__)
+  require File.expand_path('../config/environment', __dir__)
 end
 
 Teaspoon.configure do |config|
```
Comment (Polsonby) [inline spec/teaspoon_env.rb:7]: What in the heck is "teaspoon"? I just decided gem names have now officially jumped the shark 🦈

**Review by Polsonby — APPROVED** [review]
Great work 👍 thanks so much for this

File: spec/teaspoon_env.rb (hunk 4 above)
Comment (willmcb) [reply spec/teaspoon_env.rb:7]: Apparently "[Teaspoon is a Javascript test runner built for Rails.](https://github.com/jejacks0n/teaspoon)"

File: spec/features/audit_trail_spec.rb (hunk 2 above)
Comment (willmcb) [reply spec/features/audit_trail_spec.rb:11]: I believe these were things that Rubocop pulled up - I'm not sure the exact reasoning. I think it should be reworded though as it doesn't make much sense in its current form as an `it`.

File: spec/services/group_lister_spec.rb (hunk 3 above)
Comment (willmcb) [reply spec/services/group_lister_spec.rb:32]: I think this was rubocop - I can't claim to remember changing this manually!

File: spec/controllers/search_controller_spec.rb (hunk 1 above)
Comment (willmcb) [reply spec/controllers/search_controller_spec.rb:68]: Absolutely, sounds like a good approach. I'll add a TODO / ticket in the next PR 👍

File: app/mailers/problem_report_mailer.rb [hunk 5]
```diff
@@ -1,5 +1,5 @@
-class ProblemReportMailer < ActionMailer::Base
-  layout 'email'
+class ProblemReportMailer < ApplicationMailer
+  # layout 'email'
```
[the remark this answers is not in the document]
Comment (willmcb) [reply app/mailers/problem_report_mailer.rb:2]: I think it was a temporary comment - I'll remove in next PR!

File: app/helpers/people_helper.rb [hunk 6]
```diff
@@ -33,8 +33,8 @@ def team_image_tag team, options = {}
 
   def edit_person_link(name, person, options = {})
     link_to name,
-      edit_person_path(person, activity: options[:activity]),
-      options.
+            edit_person_path(person, activity: options[:activity]),
+            options.
       except(:activity).
       merge(data: edit_profile_analytics_attributes(person.id))
   end
```
[the remark this answers is not in the document]
Comment (willmcb) [reply app/helpers/people_helper.rb:40]: Yep no probs - I will refactor in upcoming PR!

**Pull request merged**